File: lib/kv.ts
/**
 * Vercel KV helpers for the TypeScript side.
 *
 * In production, uses @vercel/kv (Redis-compatible).
 * Locally or when KV is not configured, falls back to an in-memory Map.
 */

let kvModule: typeof import("@vercel/kv") | null = null;

async function getKV() {
  if (kvModule) return kvModule.kv;
  try {
    kvModule = await import("@vercel/kv");
    return kvModule.kv;
  } catch {
    return null;
  }
}

// Fallback in-memory store for local dev without Vercel KV
const memStore = new Map<string, string>();

export async function kvGet<T = unknown>(key: string): Promise<T | null> {
  const kv = await getKV();
  if (kv) return kv.get<T>(key);
  const raw = memStore.get(key);
  return raw ? (JSON.parse(raw) as T) : null;
}

export async function kvSet(key: string, value: unknown, exSeconds?: number): Promise<void> {
  const kv = await getKV();
  if (kv) {
    if (exSeconds) await kv.set(key, value, { ex: exSeconds });
    else await kv.set(key, value);
  } else {
    memStore.set(key, JSON.stringify(value));
  }
}

export async function kvDel(key: string): Promise<void> {
  const kv = await getKV();
  if (kv) await kv.del(key);
  else memStore.delete(key);
}

export async function kvKeys(pattern: string): Promise<string[]> {
  const kv = await getKV();
  if (kv) {
    // @vercel/kv supports keys with pattern
    return kv.keys(pattern);
  }
  const regex = new RegExp("^" + pattern.replace(/\*/g, ".*") + "$");
  return Array.from(memStore.keys()).filter((k) => regex.test(k));
}

export async function kvLpush(key: string, ...values: unknown[]): Promise<void> {
  const kv = await getKV();
  if (kv) {
    for (const v of values) await kv.lpush(key, v);
  } else {
    const list: unknown[] = JSON.parse(memStore.get(key) || "[]");
    list.unshift(...values);
    memStore.set(key, JSON.stringify(list));
  }
}

export async function kvLrange<T = unknown>(key: string, start: number, stop: number): Promise<T[]> {
  const kv = await getKV();
  if (kv) return kv.lrange<T>(key, start, stop);
  const list: T[] = JSON.parse(memStore.get(key) || "[]");
  return list.slice(start, stop === -1 ? undefined : stop + 1);
}

export async function kvIncr(key: string, amount = 1): Promise<number> {
  const kv = await getKV();
  if (kv) return kv.incrby(key, amount);
  const val = Number(memStore.get(key) || "0") + amount;
  memStore.set(key, String(val));
  return val;
}

File: lib_py/kv.py
"""
KV storage helper for Python serverless functions.

Uses Vercel KV (Redis-compatible REST API) in production.
Falls back to in-memory dict for local development.
"""

import json
import os
import time
import urllib.request
import urllib.parse
from typing import Any

KV_URL = os.environ.get("KV_REST_API_URL", "")
KV_TOKEN = os.environ.get("KV_REST_API_TOKEN", "")

# In-memory fallback for local dev
_mem: dict[str, Any] = {}


def _kv_request(command: list[str]) -> Any:
    """Send a command to Vercel KV REST API."""
    if not KV_URL or not KV_TOKEN:
        return None
    url = f"{KV_URL}"
    data = json.dumps(command).encode()
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "Authorization": f"Bearer {KV_TOKEN}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=10) as resp:
        result = json.loads(resp.read().decode())
    return result.get("result")


def kv_get(key: str) -> Any:
    """Get a JSON value by key."""
    if KV_URL and KV_TOKEN:
        raw = _kv_request(["GET", key])
        if raw is None:
            return None
        if isinstance(raw, str):
            try:
                return json.loads(raw)
            except (json.JSONDecodeError, TypeError):
                return raw
        return raw
    return _mem.get(key)


def kv_set(key: str, value: Any, ex: int | None = None) -> None:
    """Set a JSON value."""
    serialized = json.dumps(value)
    if KV_URL and KV_TOKEN:
        cmd = ["SET", key, serialized]
        if ex:
            cmd += ["EX", str(ex)]
        _kv_request(cmd)
    else:
        _mem[key] = value


def kv_del(key: str) -> None:
    if KV_URL and KV_TOKEN:
        _kv_request(["DEL", key])
    else:
        _mem.pop(key, None)


def kv_keys(pattern: str) -> list[str]:
    if KV_URL and KV_TOKEN:
        result = _kv_request(["KEYS", pattern])
        return result if isinstance(result, list) else []
    import re
    regex = re.compile("^" + pattern.replace("*", ".*") + "$")
    return [k for k in _mem if regex.match(k)]


def kv_lpush(key: str, *values: Any) -> None:
    if KV_URL and KV_TOKEN:
        for v in values:
            _kv_request(["LPUSH", key, json.dumps(v)])
    else:
        lst = _mem.get(key, [])
        if not isinstance(lst, list):
            lst = []
        for v in values:
            lst.insert(0, v)
        _mem[key] = lst


def kv_lrange(key: str, start: int, stop: int) -> list:
    if KV_URL and KV_TOKEN:
        result = _kv_request(["LRANGE", key, str(start), str(stop)])
        if not result:
            return []
        parsed = []
        for item in result:
            if isinstance(item, str):
                try:
                    parsed.append(json.loads(item))
                except (json.JSONDecodeError, TypeError):
                    parsed.append(item)
            else:
                parsed.append(item)
        return parsed
    lst = _mem.get(key, [])
    if not isinstance(lst, list):
        return []
    return lst[start : None if stop == -1 else stop + 1]


def kv_incr(key: str, amount: int = 1) -> int:
    if KV_URL and KV_TOKEN:
        result = _kv_request(["INCRBY", key, str(amount)])
        return int(result) if result else 0
    val = _mem.get(key, 0)
    if not isinstance(val, (int, float)):
        val = 0
    val = int(val) + amount
    _mem[key] = val
    return val
